tools/unicode_properties_parse/format_width_estimate_intervals.py

```diff
@@ -0,0 +1,181 @@
+
+        print()
+
+    # Print all ranges (right-closed), where self's width is 1 and other's width is 2.
+    def print_ranges_1_vs_2(self, other):
+        cluster_table = [False] * (self.TABLE_SIZE)
+        for u in range(self.TABLE_SIZE):
+            if (
+                self.table[u] == UnicodeWidth.IS_1
+                and other.table[u] == UnicodeWidth.IS_2
+            ):
+                cluster_table[u] = True
+
+        u = 0
+        while u < self.TABLE_SIZE:
+            if cluster_table[u]:
+                from_ = u
+                to_ = from_
+                while to_ + 1 <= self.MAX_CODE_POINT and cluster_table[to_ + 1]:
+                    to_ += 1
+                if from_ == to_:
+                    print(f"U+{from_:X}")
+                else:
+                    print(f"U+{from_:X}..U+{to_:X}")
+                u = to_
+            u += 1
+
+
+def get_table_cpp20() -> UnicodeWidthTable:
+    std_wide_ranges_cpp20 = [
+        (0x1100, 0x115F),
+        (0x2329, 0x232A),
+        (0x2E80, 0x303E),
+        (0x3040, 0xA4CF),
+        (0xAC00, 0xD7A3),
+        (0xF900, 0xFAFF),
+        (0xFE10, 0xFE19),
+        (0xFE30, 0xFE6F),
+        (0xFF00, 0xFF60),
+        (0xFFE0, 0xFFE6),
+        (0x1F300, 0x1F64F),
+        (0x1F900, 0x1F9FF),
+        (0x20000, 0x2FFFD),
+        (0x30000, 0x3FFFD),
+    ]
+
+    table = UnicodeWidthTable()
+    for rng in std_wide_ranges_cpp20:
+        table.fill_range(rng, UnicodeWidth.IS_2)
+
+    return table
+
+
+def read_from(source: TextIO) -> UnicodeWidthTable:
+    """
+    Read data from "EastAsianWidth.txt".
+    The latest version can be found at:
+    https://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt
+    The current implementation works for:
+    https://www.unicode.org/Public/15.0.0/ucd/EastAsianWidth.txt
+    To make this function work, the file should not contain a BOM.
+    """
+    table = UnicodeWidthTable()
+
+    # "The unassigned code points in the following blocks default to "W":"
+    default_wide_ranges = [
+        (0x4E00, 0x9FFF),
+        (0x3400, 0x4DBF),
+        (0xF900, 0xFAFF),
+        (0x20000, 0x2FFFD),
+        (0x30000, 0x3FFFD),
+    ]
+    for rng in default_wide_ranges:
+        table.fill_range(rng, UnicodeWidth.IS_2)
+
+    # Read explicitly assigned ranges.
+    # The lines that are not empty or pure comment are uniformly of the format "HEX(..HEX)?;(A|F|H|N|Na|W) #comment".
+    LINE_REGEX = re.compile(r"([0-9A-Z]+)(\.\.[0-9A-Z]+)?;(A|F|H|N|Na|W) *#.*")
+
+    def get_width(str: str):
+        if str == "F" or str == "W":
+            return UnicodeWidth.IS_2
+        else:
+            assert str == "A" or str == "H" or str == "N" or str == "Na"
+            return UnicodeWidth.IS_1
+
+    for line in source:
+        line = line.strip()
+        if line and not line.startswith("#"):
+            match = LINE_REGEX.fullmatch(line)
+            assert match, line # invalid line

```

Comment:
I find that the regex has been slightly outdated due to some added whitespaces in  recently-updated `EastAsianWidth-15.1.0.txt` (in https://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt))

(the regex still works for 15.0.0)

![image](https://github.com/fsb4000/STL/assets/60953653/9cae373d-36d9-4cb6-aef2-65cebecc41bc)
(Was `0000..001F;N     # Cc    [32] <control-0000>..<control-001F>
`,
|now `0000..001F     ; N  # Cc    [32] <control-0000>..<control-001F>`)